### PR TITLE
enh: add golvulncheck in linting job

### PR DESCRIPTION
**Pull request by sakshi-1505**
Towards #4459

### CODE REVIEW

**Comment by sakshi-1505** [comment]
@pellared please check

**Comment by pellared** [comment]
How have you been testing it?

**Comment by sakshi-1505** [comment]
> How have you been testing it?

I did `make lint` on my local

**Comment by sakshi-1505** [comment]
@pellared I had forgotten to push a commit which actually added govulncheck in the go modules.

File: Makefile [hunk 1]
```diff
@@ -226,6 +230,12 @@ vanity-import-check: | $(PORTO)
 misspell: | $(MISSPELL)
 	@$(MISSPELL) -w $(ALL_DOCS)
 
+.PHONY: govulncheck
+govulncheck: | $(GOVULNCHECK)
```
Comment (pellared) [inline Makefile:234]: Shouldn't it be executed for each Go module like e.g. `go-mod-tidy` or `golangci-lint`?

File: Makefile [hunk 2]
```diff
@@ -226,6 +230,12 @@ vanity-import-check: | $(PORTO)
...
+	@echo "govulncheck ./..." \
+		$(GOVULNCHECK) ./...
+
+
```
Comment (pellared) [inline Makefile:238]:
```suggestion
```

Unnecessary EOL

File: Makefile [hunk 3]
```diff
@@ -77,6 +77,10 @@ $(GOTMPL): PACKAGE=go.opentelemetry.io/build-tools/gotmpl
 GORELEASE = $(TOOLS)/gorelease
 $(GORELEASE): PACKAGE=golang.org/x/exp/cmd/gorelease
 
+GOVULNCHECK = $(TOOLS)/govulncheck
+$(TOOLS)/govulncheck: PACKAGE=golang.org/x/vuln/cmd/govulncheck
+
+
```
Comment (pellared) [inline Makefile:83]:
```suggestion
```

Unnecessary EOL

**Comment by pellared** [comment]
From https://github.com/open-telemetry/opentelemetry-go/actions/runs/6454634525/job/17521765038?pr=4598

> Working tree is not clean, did you forget to run "make precommit"?

File: Makefile (hunk 1 above)
Comment (sakshi-1505) [reply Makefile:234]: I was using `./...` which should execute the command in every folder recursively right?

Comment (sakshi-1505) [reply Makefile:234]: But let's follow the pattern which other are using, made the changes

**Comment by sakshi-1505** [comment]
@pellared I have made the changes as you suggested & ran the  precommit too now

File: Makefile (hunk 1 above)
Comment (pellared) [reply Makefile:235]: Why you will not test it? 😉 
The documentation says it works for a Go module.

File: Makefile [hunk 4]
```diff
@@ -226,6 +229,14 @@ vanity-import-check: | $(PORTO)
 misspell: | $(MISSPELL)
 	@$(MISSPELL) -w $(ALL_DOCS)
 
+.PHONY: govulncheck
+go-mod-tidy: $(OTEL_GO_MOD_DIRS:%=govulncheck/%)
+go-mod-tidy/%: DIR=$*
```
Comment (pellared) [inline Makefile:234]:
```suggestion
govulncheck: $(OTEL_GO_MOD_DIRS:%=govulncheck/%)
govulncheck/%: DIR=$*
```

File: Makefile [hunk 5]
```diff
@@ -226,6 +229,14 @@ vanity-import-check: | $(PORTO)
...
+govulncheck: | $(GOVULNCHECK)
```
Comment (pellared) [inline Makefile:235]:
```suggestion
govulncheck: $(OTEL_GO_MOD_DIRS:%=govulncheck/%)
govulncheck/%: DIR=$*
govulncheck/%: | $(GOVULNCHECK)
```

File: Makefile (hunk 1 above)
Comment (pellared) [reply Makefile:235]: For example see:

```
govulncheck ./... in ./exporters/otlp/otlpmetric/otlpmetricgrpc
Scanning your code and 298 packages across 24 dependent modules for known vulnerabilities...

Vulnerability #1: GO-2023-2043
    Improper handling of special tags within script contexts in html/template
  More info: https://pkg.go.dev/vuln/GO-2023-2043
  Standard library
    Found in: html/template@go1.21
    Fixed in: html/template@go1.21.1
    Example traces found:
      #1: internal/otest/collector.go:256:30: otest.NewHTTPCollector calls http.Server.Serve, which eventually calls template.Template.Execute
      #2: internal/otest/collector.go:256:30: otest.NewHTTPCollector calls http.Server.Serve, which eventually calls template.Template.ExecuteTemplate

Vulnerability #2: GO-2023-2041
    Improper handling of HTML-like comments in script contexts in html/template
  More info: https://pkg.go.dev/vuln/GO-2023-2041
  Standard library
    Found in: html/template@go1.21
    Fixed in: html/template@go1.21.1
    Example traces found:
      #1: internal/otest/collector.go:256:30: otest.NewHTTPCollector calls http.Server.Serve, which eventually calls template.Template.Execute
      #2: internal/otest/collector.go:256:30: otest.NewHTTPCollector calls http.Server.Serve, which eventually calls template.Template.ExecuteTemplate
```

So it needs to be called for each Go module.

**Comment by pellared** [comment]
Do you have any proposal on how to handle such issues like below?

```
govulncheck ./... in ./exporters/otlp/otlpmetric/otlpmetricgrpc
Scanning your code and 298 packages across 24 dependent modules for known vulnerabilities...

Vulnerability #1: GO-2023-2043
    Improper handling of special tags within script contexts in html/template
  More info: https://pkg.go.dev/vuln/GO-2023-2043
  Standard library
    Found in: html/template@go1.21
    Fixed in: html/template@go1.21.1
    Example traces found:
      #1: internal/otest/collector.go:256:30: otest.NewHTTPCollector calls http.Server.Serve, which eventually calls template.Template.Execute
      #2: internal/otest/collector.go:256:30: otest.NewHTTPCollector calls http.Server.Serve, which eventually calls template.Template.ExecuteTemplate

Vulnerability #2: GO-2023-2041
    Improper handling of HTML-like comments in script contexts in html/template
  More info: https://pkg.go.dev/vuln/GO-2023-2041
  Standard library
    Found in: html/template@go1.21
    Fixed in: html/template@go1.21.1
    Example traces found:
      #1: internal/otest/collector.go:256:30: otest.NewHTTPCollector calls http.Server.Serve, which eventually calls template.Template.Execute
      #2: internal/otest/collector.go:256:30: otest.NewHTTPCollector calls http.Server.Serve, which eventually calls template.Template.ExecuteTemplate
```

**Comment by pellared** [comment]
@sakshi-1505, can you create a similar PR for https://github.com/open-telemetry/opentelemetry-go-contrib?

**Review by pellared — APPROVED** [review]

**Review by dashpole — APPROVED** [review]

**Review by MrAlias — APPROVED** [review]

**Comment by pellared** [comment]
@sakshi-1505 Thank you for your contribution 🎉 

Can you create a similar PR for https://github.com/open-telemetry/opentelemetry-go-contrib and https://github.com/open-telemetry/opentelemetry-go-build-tools?

**Comment by sakshi-1505** [comment]
Sure thing!

**Pull request merged**